File: backend/internal/api/handlers/jobs.go
package handlers

import (
	"net/http"
	"strconv"
	"strings"

	"github.com/go-chi/chi/v5"

	"ytdm/backend/internal/api/response"
	"ytdm/backend/internal/apperr"
	"ytdm/backend/internal/jobs"
)

// ListJobs answers GET /jobs.
func (h *Handlers) ListJobs(w http.ResponseWriter, r *http.Request) {
	limit, offset, err := pagination(r, 50, 200)
	if err != nil {
		response.Error(w, r, err)
		return
	}

	filter := jobs.ListFilter{Limit: limit, Offset: offset}
	if raw := queryString(r, "status"); raw != "" {
		status := jobs.Status(raw)
		if !status.Valid() {
			response.Fail(w, r, apperr.CodeInvalidRequest, "The status filter is not a known job status.")
			return
		}
		filter.Status = status
	}
	if raw := queryString(r, "type"); raw != "" {
		jobType := jobs.Type(raw)
		if !jobType.Valid() {
			response.Fail(w, r, apperr.CodeInvalidRequest, "The type filter is not a known job type.")
			return
		}
		filter.Type = jobType
	}
	if raw := queryString(r, "priority"); raw != "" {
		priority := jobs.Priority(raw)
		if !priority.Valid() {
			response.Fail(w, r, apperr.CodeInvalidRequest, "The priority filter is not a known priority.")
			return
		}
		filter.Priority = priority
	}

	list, total, err := h.deps.Jobs.List(r.Context(), filter)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.List(w, list, response.Meta{Count: len(list), Total: total, Limit: limit, Offset: offset})
}

// GetJob answers GET /jobs/{id}, including the item list and the summary.
func (h *Handlers) GetJob(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	job, err := h.deps.Jobs.Get(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	items, err := h.deps.Jobs.Items(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}

	response.Data(w, http.StatusOK, map[string]any{
		"job":     job,
		"items":   items,
		"summary": job.Summary(),
	})
}

// UpdateJobRequest carries fields for PATCH /jobs/{id}.
type UpdateJobRequest struct {
	Priority *jobs.Priority `json:"priority"`
	Paused   *bool          `json:"paused"`
}

// UpdateJob answers PATCH /jobs/{id}.
func (h *Handlers) UpdateJob(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")

	var req UpdateJobRequest
	if err := decodeJSON(r, &req); err != nil {
		response.Error(w, r, err)
		return
	}

	if req.Priority == nil && req.Paused == nil {
		response.Fail(w, r, apperr.CodeInvalidRequest, "The request changes no field.")
		return
	}

	var job *jobs.Job
	var err error

	if req.Priority != nil {
		job, err = h.deps.Jobs.SetPriority(r.Context(), id, *req.Priority)
		if err != nil {
			response.Error(w, r, err)
			return
		}
	}

	if req.Paused != nil {
		if *req.Paused {
			job, err = h.deps.Jobs.Pause(r.Context(), id)
		} else {
			job, err = h.deps.Jobs.Resume(r.Context(), id)
		}
		if err != nil {
			response.Error(w, r, err)
			return
		}
	}

	response.Data(w, http.StatusOK, job)
}

// PauseJob answers POST /jobs/{id}/pause.
func (h *Handlers) PauseJob(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	job, err := h.deps.Jobs.Pause(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, job)
}

// ResumeJob answers POST /jobs/{id}/resume.
func (h *Handlers) ResumeJob(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	job, err := h.deps.Jobs.Resume(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, job)
}

// RetryFailedJob answers POST /jobs/{id}/retry-failed.
func (h *Handlers) RetryFailedJob(w http.ResponseWriter, r *http.Request) {
	id := chi.URLParam(r, "id")
	job, retried, skipped, err := h.deps.Jobs.RetryFailed(r.Context(), id)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, map[string]any{
		"job":     job,
		"retried": retried,
		"skipped": skipped,
	})
}

// RetryJobItem answers POST /jobs/{job_id}/items/{item_id}/retry.
func (h *Handlers) RetryJobItem(w http.ResponseWriter, r *http.Request) {
	jobID := chi.URLParam(r, "job_id")
	itemID := chi.URLParam(r, "item_id")

	item, err := h.deps.Jobs.RetryItem(r.Context(), jobID, itemID)
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, item)
}

// CancelJob answers DELETE /jobs/{id}. The job is cancelled, not deleted: its
// history stays readable.
func (h *Handlers) CancelJob(w http.ResponseWriter, r *http.Request) {
	job, err := h.deps.Jobs.Cancel(r.Context(), chi.URLParam(r, "id"))
	if err != nil {
		response.Error(w, r, err)
		return
	}
	response.Data(w, http.StatusOK, job)
}

// DeleteJobHistoryRequest carries options for DELETE /jobs/history.
type DeleteJobHistoryRequest struct {
	OlderThanDays int           `json:"older_than_days"`
	Statuses      []jobs.Status `json:"statuses"`
}

// DeleteJobHistory answers DELETE /jobs/history (Admin only).
func (h *Handlers) DeleteJobHistory(w http.ResponseWriter, r *http.Request) {
	days := 30
	var req DeleteJobHistoryRequest
	if r.Body != nil && r.ContentLength > 0 {
		if err := decodeJSON(r, &req); err != nil {
			response.Error(w, r, err)
			return
		}
	}
	if req.OlderThanDays > 0 {
		if req.OlderThanDays < 7 {
			response.Fail(w, r, apperr.CodeInvalidRequest, "older_than_days must be at least 7.")
			return
		}
		days = req.OlderThanDays
	}

	if raw := queryString(r, "older_than_days"); raw != "" {
		if d, err := strconv.Atoi(raw); err == nil && d >= 7 {
			days = d
		} else if err != nil || d < 7 {
			response.Fail(w, r, apperr.CodeInvalidRequest, "older_than_days must be at least 7.")
			return
		}
	}

	statuses := []jobs.Status{jobs.StatusCompleted, jobs.StatusCancelled}
	if len(req.Statuses) > 0 {
		statuses = req.Statuses
	} else if rawStatuses := queryString(r, "statuses"); rawStatuses != "" {
		parts := strings.Split(rawStatuses, ",")
		statuses = make([]jobs.Status, 0, len(parts))
		for _, p := range parts {
			s := jobs.Status(strings.TrimSpace(p))
			if s == jobs.StatusCompleted || s == jobs.StatusCancelled || s == jobs.StatusFailed {
				statuses = append(statuses, s)
			}
		}
	}

	deletedJobs, deletedItems, err := h.deps.Jobs.DeleteHistory(r.Context(), days, statuses)
	if err != nil {
		response.Error(w, r, err)
		return
	}

	response.Data(w, http.StatusOK, map[string]any{
		"deleted_jobs":  deletedJobs,
		"deleted_items": deletedItems,
	})
}

File: frontend/src/lib/api/jobs.ts
/** Download jobs: creating them, listing them, cancelling them. */

import { request, requestList } from '@/lib/api/client'
import type {
  ArtistDownloadRequest,
  Job,
  JobDetail,
  JobItem,
  JobPriority,
  JobStatus,
  JobType,
  ListMeta,
  QueueSummary,
  ReleaseDownloadRequest,
  TrackDownloadRequest,
} from '@/types/api'
import { TERMINAL_JOB_STATUSES } from '@/types/api'

/** POST /downloads/artist */
export function downloadArtist(
  body: ArtistDownloadRequest,
  signal?: AbortSignal,
): Promise<Job> {
  return request<Job>('/downloads/artist', { method: 'POST', body, signal })
}

/** POST /downloads/release */
export function downloadRelease(
  body: ReleaseDownloadRequest,
  signal?: AbortSignal,
): Promise<Job> {
  return request<Job>('/downloads/release', { method: 'POST', body, signal })
}

/** POST /downloads/track */
export function downloadTrack(
  body: TrackDownloadRequest,
  signal?: AbortSignal,
): Promise<Job> {
  return request<Job>('/downloads/track', { method: 'POST', body, signal })
}

export interface ListJobsOptions {
  status?: JobStatus
  type?: JobType
  priority?: JobPriority
  limit?: number
  offset?: number
  signal?: AbortSignal
}

export interface ListJobsResult {
  items: Job[]
  meta: ListMeta
}

/** GET /jobs */
export async function listJobs(
  options: ListJobsOptions = {},
): Promise<Job[]> {
  const result = await requestList<Job>('/jobs', {
    query: {
      status: options.status,
      type: options.type,
      priority: options.priority,
      limit: options.limit,
      offset: options.offset,
    },
    signal: options.signal,
  })
  return result.items
}

/** GET /jobs with metadata (total count for pagination) */
export function listJobsWithMeta(
  options: ListJobsOptions = {},
): Promise<ListJobsResult> {
  return requestList<Job>('/jobs', {
    query: {
      status: options.status,
      type: options.type,
      priority: options.priority,
      limit: options.limit,
      offset: options.offset,
    },
    signal: options.signal,
  })
}

/** GET /jobs/{id} — the job with its items and summary. */
export function getJob(id: string, signal?: AbortSignal): Promise<JobDetail> {
  return request<JobDetail>(`/jobs/${encodeURIComponent(id)}`, { signal })
}

/** GET /jobs/summary — live queue summary and ETA */
export function getQueueSummary(signal?: AbortSignal): Promise<QueueSummary> {
  return request<QueueSummary>('/jobs/summary', { signal })
}

/** PATCH /jobs/{id} — update job priority and/or pause state. */
export function updateJob(
  id: string,
  body: { priority?: JobPriority; paused?: boolean },
  signal?: AbortSignal,
): Promise<Job> {
  return request<Job>(`/jobs/${encodeURIComponent(id)}`, {
    method: 'PATCH',
    body,
    signal,
  })
}

/** POST /jobs/{id}/pause */
export function pauseJob(id: string, signal?: AbortSignal): Promise<Job> {
  return request<Job>(`/jobs/${encodeURIComponent(id)}/pause`, {
    method: 'POST',
    signal,
  })
}

/** POST /jobs/{id}/resume */
export function resumeJob(id: string, signal?: AbortSignal): Promise<Job> {
  return request<Job>(`/jobs/${encodeURIComponent(id)}/resume`, {
    method: 'POST',
    signal,
  })
}

/** POST /jobs/{id}/retry-failed */
export function retryFailedJob(
  id: string,
  signal?: AbortSignal,
): Promise<{ job: Job; retried: number; skipped: number }> {
  return request<{ job: Job; retried: number; skipped: number }>(
    `/jobs/${encodeURIComponent(id)}/retry-failed`,
    {
      method: 'POST',
      signal,
    },
  )
}

/** POST /jobs/{job_id}/items/{item_id}/retry */
export function retryJobItem(
  jobId: string,
  itemId: string,
  signal?: AbortSignal,
): Promise<JobItem> {
  return request<JobItem>(
    `/jobs/${encodeURIComponent(jobId)}/items/${encodeURIComponent(itemId)}/retry`,
    {
      method: 'POST',
      signal,
    },
  )
}

/** DELETE /jobs/{id} — cancels the job; its history stays readable. */
export function cancelJob(id: string, signal?: AbortSignal): Promise<Job> {
  return request<Job>(`/jobs/${encodeURIComponent(id)}`, {
    method: 'DELETE',
    signal,
  })
}

/** DELETE /jobs/history — safely purges old job history (Admin only). */
export function deleteJobHistory(
  olderThanDays?: number,
  statuses?: string[],
  signal?: AbortSignal,
): Promise<{ deleted_jobs: number; deleted_items: number }> {
  return request<{ deleted_jobs: number; deleted_items: number }>(
    '/jobs/history',
    {
      method: 'DELETE',
      query: {
        older_than_days: olderThanDays,
        statuses: statuses ? statuses.join(',') : undefined,
      },
      signal,
    },
  )
}


/* ------------------------------------------------------------------ helpers */

/** True once the job has reached a final state. */
export function isTerminal(job: Job): boolean {
  return TERMINAL_JOB_STATUSES.includes(job.status)
}

/** True while the job is queued or being worked on. */
export function isActive(job: Job): boolean {
  return !isTerminal(job)
}

/** Items that reached a final state — completed, failed or skipped. */
export function processed(job: Job): number {
  return job.completed + job.failed + job.skipped
}

/**
 * Progress in percent, or null while the total is still unknown.
 *
 * The total only exists once the catalogue has been resolved, so a job in a
 * resolving state has no meaningful percentage and must not be shown as 0 %.
 */
export function progressPercent(job: Job): number | null {
  if (job.total <= 0) return null
  return Math.min(100, Math.round((processed(job) / job.total) * 100))
}

/** The German label of a job status. */
export const JOB_STATUS_LABELS: Record<JobStatus, string> = {
  queued: 'In Warteschlange',
  resolving_artist: 'Künstler wird aufgelöst',
  resolving_releases: 'Releases werden aufgelöst',
  resolving_tracks: 'Tracks werden aufgelöst',
  deduplicating: 'Duplikate werden entfernt',
  matching: 'Quellen werden gesucht',
  downloading: 'Wird heruntergeladen',
  tagging: 'Tags werden geschrieben',
  finalizing: 'Wird finalisiert',
  retry_wait: 'Wartet auf Wiederholung',
  waiting_for_storage: 'Wartet auf Speicher',
  waiting_for_space: 'Wartet auf Speicherplatz',
  completed: 'Abgeschlossen',
  failed: 'Fehlgeschlagen',
  cancelled: 'Abgebrochen',
}

/** The German label of an item status. */
export const ITEM_STATUS_LABELS: Record<string, string> = {
  pending: 'Ausstehend',
  matching: 'Quellensuche',
  downloading: 'Download',
  tagging: 'Tagging',
  finalizing: 'Finalisierung',
  retry_wait: 'Wiederholung geplant',
  waiting_for_storage: 'Wartet auf Speicher',
  waiting_for_space: 'Wartet auf Speicherplatz',
  completed: 'Abgeschlossen',
  failed: 'Fehlgeschlagen',
  skipped: 'Übersprungen',
  cancelled: 'Abgebrochen',
}

/** The German label of a job type. */
export const JOB_TYPE_LABELS: Record<JobType, string> = {
  artist: 'Diskografie',
  release: 'Release',
  track: 'Track',
}

/** The German label of a job priority. */
export const JOB_PRIORITY_LABELS: Record<JobPriority, string> = {
  low: 'Niedrig',
  normal: 'Normal',
  high: 'Hoch',
  very_high: 'Sehr hoch',
}

/**
 * How a job counts towards the sections of the downloads page.
 *
 * A job that completed with failures is deliberately still "abgeschlossen":
 * individual track failures must not present the whole job as a failure. Only
 * a job whose own status is failed belongs in the failure section.
 */
export function section(job: Job): 'active' | 'queued' | 'done' | 'failed' {
  if (job.status === 'queued') return 'queued'
  if (job.status === 'failed') return 'failed'
  if (isTerminal(job)) return 'done'
  return 'active'
}

